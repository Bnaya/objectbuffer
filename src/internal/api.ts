import { initializeArrayBuffer, incrementRefCount } from "./store";
import { createObjectWrapper } from "./objectWrapper";
import type {
  ObjectBufferSettings,
  GlobalCarrier,
  MemoryStats,
} from "./interfaces";
import {
  arrayBufferCopyTo,
  externalArgsApiToExternalArgsApi,
  getInternalAPI,
  isSupportedTopLevelValue,
  getEndiannessOfSystem,
} from "./utils";
import {
  INITIAL_ENTRY_POINTER_TO_POINTER,
  MEM_POOL_START,
  ENDIANNESS_FLAG_POINTER,
} from "./consts";
import { OutOfMemoryError, UnsupportedOperationError } from "./exceptions";
import { saveValueIterative } from "./saveValue";
import { TransactionalAllocator } from "./TransactionalAllocator";
import { freeNoLongerUsedAddresses } from "./freeNoLongerUsedAddresses";

/**
 * https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/SharedArrayBuffer#Browser_compatibility
 *
 * Future holds additional types https://github.com/tc39/proposal-resizablearraybuffer
 */
export type ArrayBufferKind = "vanilla" | "shared";

/**
 * Create a new objectBuffer, with the given initial value
 * @param size The size of the ArrayBuffer to create (heap size)
 * @param initialValue
 * @param settings
 * @param arrayBufferConstructor
 */
export function createObjectBuffer<T = any>(
  size: number,
  initialValue: T,
  settings: ObjectBufferSettings = {},
  arrayBufferKind: ArrayBufferKind = "vanilla"
): T {
  if (!isSupportedTopLevelValue(initialValue)) {
    throw new UnsupportedOperationError();
  }

  const allocator = new TransactionalAllocator(
    {
      start: MEM_POOL_START,
      size,
    },
    arrayBufferKind === "shared"
  );

  const arrayBuffer = allocator.getArrayBuffer();

  initializeArrayBuffer(arrayBuffer);

  const carrier: GlobalCarrier = {
    allocator,
    heap: allocator.getHeap(),
  };

  const referencedPointers: number[] = [];

  allocator.transaction(() => {
    saveValueIterative(
      externalArgsApiToExternalArgsApi(settings),
      carrier,
      referencedPointers,
      INITIAL_ENTRY_POINTER_TO_POINTER,
      initialValue
    );
  });

  for (const pointer of referencedPointers) {
    incrementRefCount(carrier.heap, pointer);
  }

  const dv = new DataView(arrayBuffer);
  // endianness flag is always saved in little endian so we can read in every system endianness
  dv.setUint32(ENDIANNESS_FLAG_POINTER, getEndiannessOfSystem(), true);

  return createObjectWrapper(
    externalArgsApiToExternalArgsApi(settings),
    carrier,
    carrier.heap.u32[
      INITIAL_ENTRY_POINTER_TO_POINTER / Uint32Array.BYTES_PER_ELEMENT
    ]
  );
}

/**
 * Grow or shrink the underlying ArrayBuffer
 *
 * @unstable
 * Due to possible issues with future support of typed arrays,
 * and the upcoming proposal, this api function may be removed
 * https://github.com/tc39/proposal-resizablearraybuffer
 *
 * @param objectBuffer
 * @param newSize
 *
 */
export function unstable_resizeObjectBuffer(
  objectBuffer: unknown,
  newSize: number
) {
  if (newSize < memoryStats(objectBuffer).top) {
    throw new OutOfMemoryError();
  }

  const oldArrayBuffer = getUnderlyingArrayBuffer(objectBuffer);
  const newArrayBuffer = new ArrayBuffer(newSize);

  arrayBufferCopyTo(
    oldArrayBuffer,
    0,
    Math.min(newSize, oldArrayBuffer.byteLength),
    newArrayBuffer,
    0
  );

  unstable_replaceUnderlyingArrayBuffer(objectBuffer, newArrayBuffer);

  return newArrayBuffer;
}

export function getUnderlyingArrayBuffer(
  objectBuffer: unknown
): ArrayBuffer | SharedArrayBuffer {
  return getInternalAPI(objectBuffer).getCarrier().heap.u8.buffer;
}

/**
 * Create objectBuffer object from the given ArrayBuffer
 *
 * The given ArrayBuffer is expected to be one obtained via getUnderlyingArrayBuffer
 * This operation doesn't change any value in the ArrayBuffer
 *
 * @param settings
 * @param arrayBuffer
 */
export function loadObjectBuffer<T = any>(
  arrayBuffer: ArrayBuffer | SharedArrayBuffer,
  settings: ObjectBufferSettings = {}
): T {
  const allocator = TransactionalAllocator.load(arrayBuffer);

  const carrier: GlobalCarrier = {
    allocator,
    heap: allocator.getHeap(),
  };

  const dv = new DataView(arrayBuffer);
  // endianness flag is always saved in little endian so we can read in every system endianness
  const endiannessOfGivenAb = dv.getUint32(ENDIANNESS_FLAG_POINTER, true);

  if (endiannessOfGivenAb !== getEndiannessOfSystem()) {
    throw new Error("Endianness mismatch");
  }

  return createObjectWrapper(
    externalArgsApiToExternalArgsApi(settings),
    carrier,
    carrier.heap.u32[
      INITIAL_ENTRY_POINTER_TO_POINTER / Uint32Array.BYTES_PER_ELEMENT
    ]
  );
}

/**
 * Replace the Underlying array buffer with the given one.
 * The given ArrayBuffer is expected to be a copy of the prev ArrayBuffer,
 * just bigger or smaller (less free space)
 *
 * Consider using `resizeObjectBuffer`
 *
 * @unstable
 * Due to possible issues with future support of typed arrays,
 * and the upcoming proposal, this api function may be removed
 * https://github.com/tc39/proposal-resizablearraybuffer
 *
 * @param objectBuffer
 * @param newArrayBuffer
 */
export function unstable_replaceUnderlyingArrayBuffer(
  objectBuffer: unknown,
  newArrayBuffer: ArrayBuffer | SharedArrayBuffer
) {
  const allocator = TransactionalAllocator.load(newArrayBuffer);
  const carrier: GlobalCarrier = {
    allocator,
    heap: allocator.getHeap(),
  };

  const dv = new DataView(newArrayBuffer);
  // endianness flag is always saved in little endian so we can read in every system endianness
  const endiannessOfGivenAb = dv.getUint32(ENDIANNESS_FLAG_POINTER, true);

  if (endiannessOfGivenAb !== getEndiannessOfSystem()) {
    throw new Error("Endianness mismatch");
  }

  allocator.setNewEnd(newArrayBuffer.byteLength);

  getInternalAPI(objectBuffer).replaceCarrierContent(carrier);
}

/**
 * Return the number of free & used bytes left in the given objectBuffer
 */
export function memoryStats(objectBuffer: unknown): MemoryStats {
  const { available, total, top } = getInternalAPI(objectBuffer)
    .getCarrier()
    .allocator.stats();

  return { available, used: total - available, total, top };
}

export { reclaim, queueReclaim } from "./reclaim";

/**
 *  Update the settings of the given ObjectBuffer
 */
export function updateObjectBufferSettings(
  objectBuffer: unknown,
  options: ObjectBufferSettings
) {
  Object.assign(getInternalAPI(objectBuffer).getExternalArgs(), options);
}

export function readObjectBufferSettings(objectBuffer: unknown) {
  return getInternalAPI(objectBuffer).getExternalArgs();
}

/**
 * Free all the addresses collected using FinalizationRegistry
 * When no FinalizationRegistry/WeakRef available, use disposeWrapperObject
 * https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/FinalizationRegistry
 *
 * This is not called automatic by FinalizationRegistry,
 * because It's only safe to call it when you have a lock/similar (As any other operation)
 * And FinalizationRegistry might run when ever
 */
export function processQueuedReclaims(objectBuffer: unknown) {
  freeNoLongerUsedAddresses(getInternalAPI(objectBuffer).getCarrier());
}
