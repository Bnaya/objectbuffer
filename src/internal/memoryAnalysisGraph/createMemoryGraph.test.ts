/* eslint-env jest */

import { createObjectBuffer } from "../api";
import {
  nodesWithInternalMismatchRefCount,
  getGraphForObjectBuffer,
  jestExpectNoUseAfterFree,
} from "./memoryGraphHelpers";
describe("createMemoryGraph. snapshot graph and ensure memory", () => {
  test("basic", () => {
    const sameObject = { t: 6 };
    const ob = createObjectBuffer(1024 * 10, {
      a: 1,
      b: { t: undefined, t2: null },
      c: true,
      t: BigInt("-1"),
      ar: [1, 666, 3, 4],
      m: new Map([[1, "a"]]),
      s: new Set([6, 7, 8, 9]),
      date: new Date("1-1-2020"),
      sameObject0: sameObject,
      sameObject1: sameObject,
      sameObject2: sameObject,
      sameObject3: sameObject,
      sameObject4: sameObject,
    });

    const { graph, visitedPointers } = getGraphForObjectBuffer(ob);

    expect(graph.nodes).toMatchSnapshot("nodes");
    expect(graph.edges).toMatchSnapshot("edges");
    expect(visitedPointers).toMatchSnapshot("visitedPointers");

    jestExpectNoUseAfterFree(ob);

    expect(nodesWithInternalMismatchRefCount(graph)).toMatchInlineSnapshot(`
      Array [
        Object {
          "pointer": 64,
          "refCount": 2,
          "size": 16,
          "type": "object",
        },
      ]
    `);
  });
});
