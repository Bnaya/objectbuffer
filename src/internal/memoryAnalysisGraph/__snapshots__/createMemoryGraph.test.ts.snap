// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createMemoryGraph. snapshot graph and ensure memory basic: edges 1`] = `
Array [
  Object {
    "from": 56,
    "to": 80,
  },
  Object {
    "from": 80,
    "to": 104,
  },
  Object {
    "from": 184,
    "to": 200,
  },
  Object {
    "from": 232,
    "to": 216,
  },
  Object {
    "from": 264,
    "to": 232,
  },
  Object {
    "from": 264,
    "to": 200,
  },
  Object {
    "from": 264,
    "to": 2600,
  },
  Object {
    "from": 200,
    "to": 264,
  },
  Object {
    "from": 184,
    "to": 288,
  },
  Object {
    "from": 320,
    "to": 304,
  },
  Object {
    "from": 352,
    "to": 320,
  },
  Object {
    "from": 352,
    "to": 288,
  },
  Object {
    "from": 2352,
    "to": 2376,
  },
  Object {
    "from": 2376,
    "to": 2400,
  },
  Object {
    "from": 2440,
    "to": 2456,
  },
  Object {
    "from": 496,
    "to": 480,
  },
  Object {
    "from": 2472,
    "to": 496,
  },
  Object {
    "from": 2472,
    "to": 2456,
  },
  Object {
    "from": 2472,
    "to": 0,
  },
  Object {
    "from": 2456,
    "to": 2472,
  },
  Object {
    "from": 2440,
    "to": 2496,
  },
  Object {
    "from": 2528,
    "to": 2512,
  },
  Object {
    "from": 2560,
    "to": 2528,
  },
  Object {
    "from": 2560,
    "to": 2496,
  },
  Object {
    "from": 2560,
    "to": 1,
  },
  Object {
    "from": 2496,
    "to": 2560,
  },
  Object {
    "from": 2584,
    "to": 0,
  },
  Object {
    "from": 2440,
    "specialLabel": "linkedlist end marker",
    "to": 2584,
  },
  Object {
    "from": 2376,
    "to": 2440,
  },
  Object {
    "from": 2376,
    "to": 2472,
  },
  Object {
    "from": 2376,
    "to": 2560,
  },
  Object {
    "from": 352,
    "to": 2352,
  },
  Object {
    "from": 288,
    "to": 352,
  },
  Object {
    "from": 184,
    "to": 376,
  },
  Object {
    "from": 408,
    "to": 392,
  },
  Object {
    "from": 440,
    "to": 408,
  },
  Object {
    "from": 440,
    "to": 376,
  },
  Object {
    "from": 440,
    "to": 2,
  },
  Object {
    "from": 376,
    "to": 440,
  },
  Object {
    "from": 184,
    "to": 464,
  },
  Object {
    "from": 528,
    "to": 496,
  },
  Object {
    "from": 528,
    "to": 464,
  },
  Object {
    "from": 528,
    "to": 2328,
  },
  Object {
    "from": 464,
    "to": 528,
  },
  Object {
    "from": 184,
    "to": 552,
  },
  Object {
    "from": 584,
    "to": 568,
  },
  Object {
    "from": 616,
    "to": 584,
  },
  Object {
    "from": 616,
    "to": 552,
  },
  Object {
    "from": 2176,
    "to": 2208,
  },
  Object {
    "from": 2208,
    "to": 2304,
  },
  Object {
    "from": 2208,
    "to": 2280,
  },
  Object {
    "from": 2208,
    "to": 2256,
  },
  Object {
    "from": 2208,
    "to": 2232,
  },
  Object {
    "from": 616,
    "to": 2176,
  },
  Object {
    "from": 552,
    "to": 616,
  },
  Object {
    "from": 184,
    "to": 640,
  },
  Object {
    "from": 672,
    "to": 656,
  },
  Object {
    "from": 704,
    "to": 672,
  },
  Object {
    "from": 704,
    "to": 640,
  },
  Object {
    "from": 1992,
    "to": 2016,
  },
  Object {
    "from": 2016,
    "to": 2040,
  },
  Object {
    "from": 2080,
    "to": 2096,
  },
  Object {
    "from": 2136,
    "to": 2112,
  },
  Object {
    "from": 2136,
    "to": 2096,
  },
  Object {
    "from": 2136,
    "to": 232,
  },
  Object {
    "from": 2096,
    "to": 2136,
  },
  Object {
    "from": 2160,
    "to": 0,
  },
  Object {
    "from": 2080,
    "specialLabel": "linkedlist end marker",
    "to": 2160,
  },
  Object {
    "from": 2016,
    "to": 2080,
  },
  Object {
    "from": 2016,
    "to": 2136,
  },
  Object {
    "from": 704,
    "to": 1992,
  },
  Object {
    "from": 640,
    "to": 704,
  },
  Object {
    "from": 184,
    "to": 728,
  },
  Object {
    "from": 760,
    "to": 744,
  },
  Object {
    "from": 792,
    "to": 760,
  },
  Object {
    "from": 792,
    "to": 728,
  },
  Object {
    "from": 1616,
    "to": 1640,
  },
  Object {
    "from": 1640,
    "to": 1664,
  },
  Object {
    "from": 1704,
    "to": 1720,
  },
  Object {
    "from": 1760,
    "to": 1736,
  },
  Object {
    "from": 1760,
    "to": 1720,
  },
  Object {
    "from": 1760,
    "to": 0,
  },
  Object {
    "from": 1720,
    "to": 1760,
  },
  Object {
    "from": 1704,
    "to": 1784,
  },
  Object {
    "from": 1824,
    "to": 1800,
  },
  Object {
    "from": 1824,
    "to": 1784,
  },
  Object {
    "from": 1824,
    "to": 0,
  },
  Object {
    "from": 1784,
    "to": 1824,
  },
  Object {
    "from": 1704,
    "to": 1848,
  },
  Object {
    "from": 1888,
    "to": 1864,
  },
  Object {
    "from": 1888,
    "to": 1848,
  },
  Object {
    "from": 1888,
    "to": 0,
  },
  Object {
    "from": 1848,
    "to": 1888,
  },
  Object {
    "from": 1704,
    "to": 1912,
  },
  Object {
    "from": 1952,
    "to": 1928,
  },
  Object {
    "from": 1952,
    "to": 1912,
  },
  Object {
    "from": 1952,
    "to": 0,
  },
  Object {
    "from": 1912,
    "to": 1952,
  },
  Object {
    "from": 1976,
    "to": 0,
  },
  Object {
    "from": 1704,
    "specialLabel": "linkedlist end marker",
    "to": 1976,
  },
  Object {
    "from": 1640,
    "to": 1704,
  },
  Object {
    "from": 1640,
    "to": 1760,
  },
  Object {
    "from": 1640,
    "to": 1824,
  },
  Object {
    "from": 1640,
    "to": 1888,
  },
  Object {
    "from": 1640,
    "to": 1952,
  },
  Object {
    "from": 792,
    "to": 1616,
  },
  Object {
    "from": 728,
    "to": 792,
  },
  Object {
    "from": 184,
    "to": 816,
  },
  Object {
    "from": 848,
    "to": 832,
  },
  Object {
    "from": 880,
    "to": 848,
  },
  Object {
    "from": 880,
    "to": 816,
  },
  Object {
    "from": 880,
    "to": 1584,
  },
  Object {
    "from": 816,
    "to": 880,
  },
  Object {
    "from": 184,
    "to": 904,
  },
  Object {
    "from": 944,
    "to": 920,
  },
  Object {
    "from": 976,
    "to": 944,
  },
  Object {
    "from": 976,
    "to": 904,
  },
  Object {
    "from": 1400,
    "to": 1424,
  },
  Object {
    "from": 1424,
    "to": 1448,
  },
  Object {
    "from": 1488,
    "to": 1504,
  },
  Object {
    "from": 1520,
    "to": 496,
  },
  Object {
    "from": 1520,
    "to": 1504,
  },
  Object {
    "from": 1520,
    "to": 1560,
  },
  Object {
    "from": 1504,
    "to": 1520,
  },
  Object {
    "from": 1544,
    "to": 0,
  },
  Object {
    "from": 1488,
    "specialLabel": "linkedlist end marker",
    "to": 1544,
  },
  Object {
    "from": 1424,
    "to": 1488,
  },
  Object {
    "from": 1424,
    "to": 1520,
  },
  Object {
    "from": 976,
    "to": 1400,
  },
  Object {
    "from": 904,
    "to": 976,
  },
  Object {
    "from": 184,
    "to": 1000,
  },
  Object {
    "from": 1040,
    "to": 1016,
  },
  Object {
    "from": 1072,
    "to": 1040,
  },
  Object {
    "from": 1072,
    "to": 1000,
  },
  Object {
    "from": 1072,
    "to": 1400,
  },
  Object {
    "from": 1000,
    "to": 1072,
  },
  Object {
    "from": 184,
    "to": 1096,
  },
  Object {
    "from": 1136,
    "to": 1112,
  },
  Object {
    "from": 1168,
    "to": 1136,
  },
  Object {
    "from": 1168,
    "to": 1096,
  },
  Object {
    "from": 1168,
    "to": 1400,
  },
  Object {
    "from": 1096,
    "to": 1168,
  },
  Object {
    "from": 184,
    "to": 1192,
  },
  Object {
    "from": 1232,
    "to": 1208,
  },
  Object {
    "from": 1264,
    "to": 1232,
  },
  Object {
    "from": 1264,
    "to": 1192,
  },
  Object {
    "from": 1264,
    "to": 1400,
  },
  Object {
    "from": 1192,
    "to": 1264,
  },
  Object {
    "from": 184,
    "to": 1288,
  },
  Object {
    "from": 1328,
    "to": 1304,
  },
  Object {
    "from": 1360,
    "to": 1328,
  },
  Object {
    "from": 1360,
    "to": 1288,
  },
  Object {
    "from": 1360,
    "to": 1400,
  },
  Object {
    "from": 1288,
    "to": 1360,
  },
  Object {
    "from": 1384,
    "to": 0,
  },
  Object {
    "from": 184,
    "specialLabel": "linkedlist end marker",
    "to": 1384,
  },
  Object {
    "from": 80,
    "to": 184,
  },
  Object {
    "from": 80,
    "to": 264,
  },
  Object {
    "from": 80,
    "to": 352,
  },
  Object {
    "from": 80,
    "to": 440,
  },
  Object {
    "from": 80,
    "to": 528,
  },
  Object {
    "from": 80,
    "to": 616,
  },
  Object {
    "from": 80,
    "to": 704,
  },
  Object {
    "from": 80,
    "to": 792,
  },
  Object {
    "from": 80,
    "to": 880,
  },
  Object {
    "from": 80,
    "to": 976,
  },
  Object {
    "from": 80,
    "to": 1072,
  },
  Object {
    "from": 80,
    "to": 1168,
  },
  Object {
    "from": 80,
    "to": 1264,
  },
  Object {
    "from": 80,
    "to": 1360,
  },
]
`;

exports[`createMemoryGraph. snapshot graph and ensure memory basic: nodes 1`] = `
Array [
  Object {
    "pointer": 56,
    "refCount": 2,
    "size": 16,
    "type": "object",
  },
  Object {
    "pointer": 80,
    "refCount": undefined,
    "size": 14,
    "type": "hashmap",
  },
  Object {
    "pointer": 104,
    "refCount": undefined,
    "size": 68,
    "type": "hashmapBuckets",
  },
  Object {
    "pointer": 184,
    "refCount": undefined,
    "size": 8,
    "type": "linkedList",
  },
  Object {
    "pointer": 200,
    "refCount": undefined,
    "size": 8,
    "type": "linkedListItem",
  },
  Object {
    "pointer": 264,
    "refCount": undefined,
    "size": 16,
    "type": "hashmapNode",
  },
  Object {
    "pointer": 232,
    "refCount": 2,
    "size": 20,
    "type": "string",
  },
  Object {
    "pointer": 216,
    "refCount": undefined,
    "size": 1,
    "type": "stringData",
  },
  Object {
    "pointer": 2600,
    "refCount": undefined,
    "size": 16,
    "type": "number",
  },
  Object {
    "pointer": 288,
    "refCount": undefined,
    "size": 8,
    "type": "linkedListItem",
  },
  Object {
    "pointer": 352,
    "refCount": undefined,
    "size": 16,
    "type": "hashmapNode",
  },
  Object {
    "pointer": 320,
    "refCount": 1,
    "size": 20,
    "type": "string",
  },
  Object {
    "pointer": 304,
    "refCount": undefined,
    "size": 1,
    "type": "stringData",
  },
  Object {
    "pointer": 2352,
    "refCount": 1,
    "size": 16,
    "type": "object",
  },
  Object {
    "pointer": 2376,
    "refCount": undefined,
    "size": 14,
    "type": "hashmap",
  },
  Object {
    "pointer": 2400,
    "refCount": undefined,
    "size": 32,
    "type": "hashmapBuckets",
  },
  Object {
    "pointer": 2440,
    "refCount": undefined,
    "size": 8,
    "type": "linkedList",
  },
  Object {
    "pointer": 2456,
    "refCount": undefined,
    "size": 8,
    "type": "linkedListItem",
  },
  Object {
    "pointer": 2472,
    "refCount": undefined,
    "size": 16,
    "type": "hashmapNode",
  },
  Object {
    "pointer": 496,
    "refCount": 3,
    "size": 20,
    "type": "string",
  },
  Object {
    "pointer": 480,
    "refCount": undefined,
    "size": 1,
    "type": "stringData",
  },
  Object {
    "pointer": 2496,
    "refCount": undefined,
    "size": 8,
    "type": "linkedListItem",
  },
  Object {
    "pointer": 2560,
    "refCount": undefined,
    "size": 16,
    "type": "hashmapNode",
  },
  Object {
    "pointer": 2528,
    "refCount": 1,
    "size": 20,
    "type": "string",
  },
  Object {
    "pointer": 2512,
    "refCount": undefined,
    "size": 2,
    "type": "stringData",
  },
  Object {
    "pointer": 2584,
    "refCount": undefined,
    "size": 8,
    "type": "linkedListItem",
  },
  Object {
    "pointer": 376,
    "refCount": undefined,
    "size": 8,
    "type": "linkedListItem",
  },
  Object {
    "pointer": 440,
    "refCount": undefined,
    "size": 16,
    "type": "hashmapNode",
  },
  Object {
    "pointer": 408,
    "refCount": 1,
    "size": 20,
    "type": "string",
  },
  Object {
    "pointer": 392,
    "refCount": undefined,
    "size": 1,
    "type": "stringData",
  },
  Object {
    "pointer": 464,
    "refCount": undefined,
    "size": 8,
    "type": "linkedListItem",
  },
  Object {
    "pointer": 528,
    "refCount": undefined,
    "size": 16,
    "type": "hashmapNode",
  },
  Object {
    "pointer": 2328,
    "refCount": undefined,
    "size": 16,
    "type": "bigintNegative",
  },
  Object {
    "pointer": 552,
    "refCount": undefined,
    "size": 8,
    "type": "linkedListItem",
  },
  Object {
    "pointer": 616,
    "refCount": undefined,
    "size": 16,
    "type": "hashmapNode",
  },
  Object {
    "pointer": 584,
    "refCount": 1,
    "size": 20,
    "type": "string",
  },
  Object {
    "pointer": 568,
    "refCount": undefined,
    "size": 2,
    "type": "stringData",
  },
  Object {
    "pointer": 2176,
    "refCount": 1,
    "size": 24,
    "type": "array",
  },
  Object {
    "pointer": 2208,
    "refCount": undefined,
    "size": 16,
    "type": "arrayPointers",
  },
  Object {
    "pointer": 2304,
    "refCount": undefined,
    "size": 16,
    "type": "number",
  },
  Object {
    "pointer": 2280,
    "refCount": undefined,
    "size": 16,
    "type": "number",
  },
  Object {
    "pointer": 2256,
    "refCount": undefined,
    "size": 16,
    "type": "number",
  },
  Object {
    "pointer": 2232,
    "refCount": undefined,
    "size": 16,
    "type": "number",
  },
  Object {
    "pointer": 640,
    "refCount": undefined,
    "size": 8,
    "type": "linkedListItem",
  },
  Object {
    "pointer": 704,
    "refCount": undefined,
    "size": 16,
    "type": "hashmapNode",
  },
  Object {
    "pointer": 672,
    "refCount": 1,
    "size": 20,
    "type": "string",
  },
  Object {
    "pointer": 656,
    "refCount": undefined,
    "size": 1,
    "type": "stringData",
  },
  Object {
    "pointer": 1992,
    "refCount": 1,
    "size": 16,
    "type": "map",
  },
  Object {
    "pointer": 2016,
    "refCount": undefined,
    "size": 14,
    "type": "hashmap",
  },
  Object {
    "pointer": 2040,
    "refCount": undefined,
    "size": 32,
    "type": "hashmapBuckets",
  },
  Object {
    "pointer": 2080,
    "refCount": undefined,
    "size": 8,
    "type": "linkedList",
  },
  Object {
    "pointer": 2096,
    "refCount": undefined,
    "size": 8,
    "type": "linkedListItem",
  },
  Object {
    "pointer": 2136,
    "refCount": undefined,
    "size": 16,
    "type": "hashmapNode",
  },
  Object {
    "pointer": 2112,
    "refCount": undefined,
    "size": 16,
    "type": "number",
  },
  Object {
    "pointer": 2160,
    "refCount": undefined,
    "size": 8,
    "type": "linkedListItem",
  },
  Object {
    "pointer": 728,
    "refCount": undefined,
    "size": 8,
    "type": "linkedListItem",
  },
  Object {
    "pointer": 792,
    "refCount": undefined,
    "size": 16,
    "type": "hashmapNode",
  },
  Object {
    "pointer": 760,
    "refCount": 1,
    "size": 20,
    "type": "string",
  },
  Object {
    "pointer": 744,
    "refCount": undefined,
    "size": 1,
    "type": "stringData",
  },
  Object {
    "pointer": 1616,
    "refCount": 1,
    "size": 16,
    "type": "set",
  },
  Object {
    "pointer": 1640,
    "refCount": undefined,
    "size": 14,
    "type": "hashmap",
  },
  Object {
    "pointer": 1664,
    "refCount": undefined,
    "size": 32,
    "type": "hashmapBuckets",
  },
  Object {
    "pointer": 1704,
    "refCount": undefined,
    "size": 8,
    "type": "linkedList",
  },
  Object {
    "pointer": 1720,
    "refCount": undefined,
    "size": 8,
    "type": "linkedListItem",
  },
  Object {
    "pointer": 1760,
    "refCount": undefined,
    "size": 16,
    "type": "hashmapNode",
  },
  Object {
    "pointer": 1736,
    "refCount": undefined,
    "size": 16,
    "type": "number",
  },
  Object {
    "pointer": 1784,
    "refCount": undefined,
    "size": 8,
    "type": "linkedListItem",
  },
  Object {
    "pointer": 1824,
    "refCount": undefined,
    "size": 16,
    "type": "hashmapNode",
  },
  Object {
    "pointer": 1800,
    "refCount": undefined,
    "size": 16,
    "type": "number",
  },
  Object {
    "pointer": 1848,
    "refCount": undefined,
    "size": 8,
    "type": "linkedListItem",
  },
  Object {
    "pointer": 1888,
    "refCount": undefined,
    "size": 16,
    "type": "hashmapNode",
  },
  Object {
    "pointer": 1864,
    "refCount": undefined,
    "size": 16,
    "type": "number",
  },
  Object {
    "pointer": 1912,
    "refCount": undefined,
    "size": 8,
    "type": "linkedListItem",
  },
  Object {
    "pointer": 1952,
    "refCount": undefined,
    "size": 16,
    "type": "hashmapNode",
  },
  Object {
    "pointer": 1928,
    "refCount": undefined,
    "size": 16,
    "type": "number",
  },
  Object {
    "pointer": 1976,
    "refCount": undefined,
    "size": 8,
    "type": "linkedListItem",
  },
  Object {
    "pointer": 816,
    "refCount": undefined,
    "size": 8,
    "type": "linkedListItem",
  },
  Object {
    "pointer": 880,
    "refCount": undefined,
    "size": 16,
    "type": "hashmapNode",
  },
  Object {
    "pointer": 848,
    "refCount": 1,
    "size": 20,
    "type": "string",
  },
  Object {
    "pointer": 832,
    "refCount": undefined,
    "size": 4,
    "type": "stringData",
  },
  Object {
    "pointer": 1584,
    "refCount": 1,
    "size": 16,
    "type": "date",
  },
  Object {
    "pointer": 904,
    "refCount": undefined,
    "size": 8,
    "type": "linkedListItem",
  },
  Object {
    "pointer": 976,
    "refCount": undefined,
    "size": 16,
    "type": "hashmapNode",
  },
  Object {
    "pointer": 944,
    "refCount": 1,
    "size": 20,
    "type": "string",
  },
  Object {
    "pointer": 920,
    "refCount": undefined,
    "size": 11,
    "type": "stringData",
  },
  Object {
    "pointer": 1400,
    "refCount": 5,
    "size": 16,
    "type": "object",
  },
  Object {
    "pointer": 1424,
    "refCount": undefined,
    "size": 14,
    "type": "hashmap",
  },
  Object {
    "pointer": 1448,
    "refCount": undefined,
    "size": 32,
    "type": "hashmapBuckets",
  },
  Object {
    "pointer": 1488,
    "refCount": undefined,
    "size": 8,
    "type": "linkedList",
  },
  Object {
    "pointer": 1504,
    "refCount": undefined,
    "size": 8,
    "type": "linkedListItem",
  },
  Object {
    "pointer": 1520,
    "refCount": undefined,
    "size": 16,
    "type": "hashmapNode",
  },
  Object {
    "pointer": 1560,
    "refCount": undefined,
    "size": 16,
    "type": "number",
  },
  Object {
    "pointer": 1544,
    "refCount": undefined,
    "size": 8,
    "type": "linkedListItem",
  },
  Object {
    "pointer": 1000,
    "refCount": undefined,
    "size": 8,
    "type": "linkedListItem",
  },
  Object {
    "pointer": 1072,
    "refCount": undefined,
    "size": 16,
    "type": "hashmapNode",
  },
  Object {
    "pointer": 1040,
    "refCount": 1,
    "size": 20,
    "type": "string",
  },
  Object {
    "pointer": 1016,
    "refCount": undefined,
    "size": 11,
    "type": "stringData",
  },
  Object {
    "pointer": 1096,
    "refCount": undefined,
    "size": 8,
    "type": "linkedListItem",
  },
  Object {
    "pointer": 1168,
    "refCount": undefined,
    "size": 16,
    "type": "hashmapNode",
  },
  Object {
    "pointer": 1136,
    "refCount": 1,
    "size": 20,
    "type": "string",
  },
  Object {
    "pointer": 1112,
    "refCount": undefined,
    "size": 11,
    "type": "stringData",
  },
  Object {
    "pointer": 1192,
    "refCount": undefined,
    "size": 8,
    "type": "linkedListItem",
  },
  Object {
    "pointer": 1264,
    "refCount": undefined,
    "size": 16,
    "type": "hashmapNode",
  },
  Object {
    "pointer": 1232,
    "refCount": 1,
    "size": 20,
    "type": "string",
  },
  Object {
    "pointer": 1208,
    "refCount": undefined,
    "size": 11,
    "type": "stringData",
  },
  Object {
    "pointer": 1288,
    "refCount": undefined,
    "size": 8,
    "type": "linkedListItem",
  },
  Object {
    "pointer": 1360,
    "refCount": undefined,
    "size": 16,
    "type": "hashmapNode",
  },
  Object {
    "pointer": 1328,
    "refCount": 1,
    "size": 20,
    "type": "string",
  },
  Object {
    "pointer": 1304,
    "refCount": undefined,
    "size": 11,
    "type": "stringData",
  },
  Object {
    "pointer": 1384,
    "refCount": undefined,
    "size": 8,
    "type": "linkedListItem",
  },
]
`;

exports[`createMemoryGraph. snapshot graph and ensure memory basic: visitedPointers 1`] = `
Set {
  56,
  80,
  104,
  184,
  200,
  264,
  232,
  216,
  2600,
  288,
  352,
  320,
  304,
  2352,
  2376,
  2400,
  2440,
  2456,
  2472,
  496,
  480,
  2496,
  2560,
  2528,
  2512,
  2584,
  376,
  440,
  408,
  392,
  464,
  528,
  2328,
  552,
  616,
  584,
  568,
  2176,
  2208,
  2304,
  2280,
  2256,
  2232,
  640,
  704,
  672,
  656,
  1992,
  2016,
  2040,
  2080,
  2096,
  2136,
  2112,
  2160,
  728,
  792,
  760,
  744,
  1616,
  1640,
  1664,
  1704,
  1720,
  1760,
  1736,
  1784,
  1824,
  1800,
  1848,
  1888,
  1864,
  1912,
  1952,
  1928,
  1976,
  816,
  880,
  848,
  832,
  1584,
  904,
  976,
  944,
  920,
  1400,
  1424,
  1448,
  1488,
  1504,
  1520,
  1560,
  1544,
  1000,
  1072,
  1040,
  1016,
  1096,
  1168,
  1136,
  1112,
  1192,
  1264,
  1232,
  1208,
  1288,
  1360,
  1328,
  1304,
  1384,
}
`;
